### PR TITLE
Fix diagnostic suppressor feature for compiler warnings in presence o…

### LINKED ISSUE #36215: Fix diagnostic suppressor feature for compiler warnings in presence of /warnaserror
Tracks the work reverted from https://github.com/dotnet/roslyn/pull/36067 as par of https://github.com/dotnet/roslyn/pull/36067#discussion_r291283127.

Following unit tests should be unskipped as part of this work:
1. C#:  `TestSuppression_CompilerParserWarningAsError`
2. VB: `TestSuppression_CompilerWarningAsError`

### PR DESCRIPTION
**Pull request by mavasani**
…f /warnaserror

Follow up to #36067
Fixes #36215

### CODE REVIEW

**Comment by AlekseyTs** [comment]
>         /// below error severity.

It looks like this portion of the comment should be adjusted as well.

---
Refers to: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:568 in a1859bf. [](commit_id = a1859bf893f8f7427db8f9d6ec6e505d51a36666, deletion_comment = False)

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 1]
```diff
@@ -571,22 +571,14 @@ internal static bool HasUnsuppressedErrors(DiagnosticBag diagnostics)
         {
             foreach (var diag in diagnostics.AsEnumerable())
             {
-                if (IsReportedError(diag))
+                if (diag.DefaultSeverity == DiagnosticSeverity.Error && !diag.IsSuppressed)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: >!diag.IsSuppressed [](start = 72, length = 18)

Do we need to check this condition? Can "errors by default" be suppressed?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 2]
```diff
@@ -1347,7 +1347,7 @@ internal void CompleteCompilationEventQueue_NoLock()
         /// </summary>
         /// <param name="accumulator">Bag to which filtered diagnostics will be added.</param>
         /// <param name="incoming">Diagnostics to be filtered.</param>
-        /// <returns>True if there were no errors or warnings-as-errors.</returns>
+        /// <returns>True if there were no errors.</returns>
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: >True if there were no errors. [](start = 21, length = 29)

Consider adjusting as well: "No diagnostic with default severity ..."

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 3]
```diff
@@ -1378,7 +1378,7 @@ internal bool FilterAndAppendDiagnostics(DiagnosticBag accumulator, IEnumerable<
                 {
                     continue;
                 }
-                else if (filtered.Severity == DiagnosticSeverity.Error &&
+                else if (filtered.DefaultSeverity == DiagnosticSeverity.Error &&
                     !filtered.IsSuppressed)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: >!filtered.IsSuppressed [](start = 20, length = 22)

Do we still have to check this?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 1 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: Yes, analyzer diagnostics which are errors by default can be suppressed in source by pragma/suppress message attributes. They cannot be programmatically suppressed as per the finalized design of DiagnosticSuppressor feature.

---
In reply to: [295479590](https://github.com/dotnet/roslyn/pull/36582#discussion_r295479590) [](ancestors = 295479590)

Comment (mavasani) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: They can be suppressed even with /nowarn. See [http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests/CommandLineTests.cs,9967](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests/CommandLineTests.cs,9967)

---
In reply to: [295495458](https://github.com/dotnet/roslyn/pull/36582#discussion_r295495458) [](ancestors = 295495458,295479590)

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 3 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs]: Yes, for the same reason as in the above comment.

---
In reply to: [295480408](https://github.com/dotnet/roslyn/pull/36582#discussion_r295480408) [](ancestors = 295480408)

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 1 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: I am actually going to improve this check to `if (diag.DefaultSeverity == DiagnosticSeverity.Error && diag.IsNotConfigurable())`, which makes the semantics clear - return true on any error diagnostic that cannot be configured/suppressed. This includes all compiler diagnostics and analyzer errors that are guaranteed to break build.

---
In reply to: [295497929](https://github.com/dotnet/roslyn/pull/36582#discussion_r295497929) [](ancestors = 295497929,295495458,295479590)

File: src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs [hunk 4]
```diff
@@ -559,6 +559,14 @@ internal virtual bool IsNotConfigurable()
         {
             return AnalyzerManager.HasNotConfigurableTag(this.CustomTags);
         }
+
+        /// <summary>
+        /// Returns true if this is an error diagnostic which cannot be suppressed and is guaranteed to break the build.
+        /// Only diagnostics which have default severity error and are tagged as NotConfigurable fall in this bucket.
+        /// This includes all compiler error diagnostics and specific analyzer error diagnostics that are marked as not configurable by the analyzer author.
+        /// </summary>
+        internal bool IsUnsuppressedError()
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: >IsUnsuppressedError [](start = 22, length = 19)

IsUnsuppressableError ?

Comment (mavasani) [reply src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: Yes, that sounds like a better name. I will update.

**Pull request merged**